include/envoy/ssl/handshaker.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Ssl {
+
+class HandshakerCallbacks {
+public:
+  virtual ~HandshakerCallbacks() = default;
+
+  /**
+   * Hands off the internally-held SSL object for external manipulation.
+   */
+  virtual bssl::UniquePtr<SSL> HandOff() PURE;
+  /**
+   * Accepts an SSL object for internal storage.
+   */
+  virtual void HandBack(bssl::UniquePtr<SSL> ssl) PURE;
+  /**
+   * Called when a handshake is successfully performed.
+   */
+  virtual void OnSuccessCb(SSL* ssl) PURE;
+  /**
+   * Called when a handshake fails.
+   */
+  virtual void OnFailureCb() PURE;
+};
+
+class Handshaker {
+public:
+  virtual ~Handshaker() = default;
+
+  // The initialize method must be called before
+  // doHandshake() is called.
+  virtual void initialize(SSL& ssl) PURE;
+
+  /**
+   * Do the handshake.
+   *
+   *  * |state| is a mutable reference.
+   *  * |ssl| is a pointer, and this method may be called with |ssl| == nullptr,
```

Comment:
What does the handshaker do if `ssl` is null here? Should the call be omitted by the caller in that case?

Reply:
Clarified guidance. I don't think the call should be omitted; doHandshake() works like SSL_do_handshake() and should get called repeatedly until done, including during the period in which `|ssl| == nullptr`.